lib/spack/spack/binary_distribution.py

```diff
@@ -440,22 +445,24 @@ def get_specs():
             for link in links:
                 if re.search("spec.yaml", link) and re.search(path, link):
                     urls.add(link)
-        for link in urls:
-            with Stage(link, name="build_cache", keep=True) as stage:
-                try:
-                    stage.fetch()
-                except fs.FetchError:
-                    continue
-                with open(stage.save_filename, 'r') as f:
-                    # read the spec from the build cache file. All specs
-                    # in build caches are concrete (as they aer built) so
-                    # we need to mark this spec concrete on read-in.
-                    spec = spack.spec.Spec.from_yaml(f)
-                    spec._mark_concrete()
-
-                    specs.add(spec)
-                    durls[spec].append(link)
-    return specs, durls
+
+    durls = defaultdict(list)
+    for link in urls:
+        with Stage(link, name="build_cache", keep=True) as stage:
+            try:
+                stage.fetch()
+            except fs.FetchError:
+                continue
+            with open(stage.save_filename, 'r') as f:
+                # read the spec from the build cache file. All specs
+                # in build caches are concrete (as they are built) so
+                # we need to mark this spec concrete on read-in.
+                spec = spack.spec.Spec.from_yaml(f)
+                spec._mark_concrete()
+                durls[spec].append(link)
+
+    spack.binary_cache_retrieved_specs = durls
+    return durls
```

Comment:
Are we returning a list of urls from the get_specs() method here?

Reply:
This is a dictionary of `spack spec` to `list of download links where that spec can be retrieved`. 

Originally this function returned a set of specs and the above dictionary. I thought that redundant (the keyset of the dictionary contains the former) so I removed that return value. Now that I look over it again I don't think the links are used anywhere (e.g. `binary_distribution.download_tarball` takes a spec and checks all the mirrors again).

@gartung FYI in case you have any opinions on this

Reply:
I added the check for the existence of spec.yaml because the stage.fetch method would print a warning that it had already been downloaded. I added the ability to force the download of the spec.yaml files.

Reply:
I think at some point I was planning to allow the selection of which mirror to download from.

Reply:
https://github.com/scheibelp/spack/blob/43c29bb88290b2df6f3228be70c1a58839a94a55/lib/spack/spack/cmd/buildcache.py#L165
Will this need to be changed to use specs.keys?

Reply:
It will need to be changed here as wellhttps://github.com/scheibelp/spack/blob/43c29bb88290b2df6f3228be70c1a58839a94a55/lib/spack/spack/cmd/buildcache.py#L308

Reply:
It would be easier to remove durls and return specs instead.

Reply:
That said, I would keep the dictionary and use 
for spec in sorted(specs.keys)
I might want to add a preferred mirror option later.

Reply:
It is OK with me to remove durls and return specs instead.

Reply:
> Could we just return the list of specs here?

Yes (sounds like we agree that is simpler and works fine). I'll aim to do that before noon tomorrow.

Reply:
https://github.com/scheibelp/spack/blob/43c29bb88290b2df6f3228be70c1a58839a94a55/lib/spack/spack/binary_distribution.py#L248
I need to add the keyword
symlinks=True to the install_tree call
otherwise I get copies of files instead of symlinks.

Reply:
@scheibelp can you add that or should I submit a pull request?

Reply:
> I need to add the keyword `symlinks=True` to the `install_tree` call otherwise I get copies of files instead of symlinks.

I'm a bit confused - is this a potential issue with this PR? Why do you want symlinks instead of file copies (if this is intended to be packaged up and copied to other systems)?